src/System.Data.SqlClient/src/System/Data/SqlClient/TdsParserHelperClasses.cs

```diff
@@ -280,20 +280,32 @@ sealed internal class SqlLoginAck
 
     sealed internal class _SqlMetaData : SqlMetaDataPriv
     {
+        [Flags]
+        private enum _SqlMetadataFlags : int
```

Comment:
The main perf improvement is because you packed the many different booleans into a single flag, is that right?  
7 booleans are now packed into a Single Int causing an improvement in allocations?

Reply:
For `_SqlMetaData` yes. sharplab.io showed that they were represented as a byte each so we get back an int which isn't great but the number of allocations makes it more significant than you'd think.

The base type changes remove some reference sized fields and make others optional inside child classes so there's net gain there as well.

Reply:
Why not change this to byte instead of using an int? The flag won't need more than 255 values.
```suggestion
        private enum _SqlMetadataFlags : byte
```

Reply:
> Because if it's a byte every use of it will have to sign extend up to an int when it's masked which costs cycles. 

I don't follow this explanation.

Reply:
The layout of the fields from the base class and main class mean that the runtime layout of 5 bools is 5 bytes in length, if you change that to be bit fields in a byte you take off 4 bytes. The structure in memory is going to be 32 bit aligned though so 5 bytes or 1 means you end up with something smaller but every time you access that last byte the runtime is going to translate it in the runtime to an int to work on it. 

Looking at that I thought that there was no point using a byte when it'll consume the same space as an int and make the runtime do extra work to deal with it. I'm not that attached to it so if you prefer a byte that's fine I just think the hidden cost isn't worth it at the moment and if we later need more fields we can change it since it's internal.

Reply:
Have a look at this: https://gist.github.com/Wraith2/2cea2a23cb26ff082a7cdb42a82d048a on sharplab.io in release and run mode.

If you scroll to the right hand side you can see that tableNum op operand are packed into an int. flags is put into the next int and there are 3 empty bytes at the end because the runtime wants to align things at 32 bit binaries. If we type the enum as byte it has to unpack it from that int at the end every time it wants to load or store it, if it's typed as a byte it just loads/stores with no extra steps.

Reply:
Got it . Thanks for explaining. For some reason I am not receiving emails of your comments and updates on this PR and for that matter, I am not getting any emails on being tagged either. Hence the delay.

At this point the changes look good to me. However I am not sure if the metadata changes have a test. Did you get a chance to confirm this? IIRC the _SqlMetadata should be surfaced in the SqlDataReader.GetSchemaTable() API

Reply:
There are no tests which specifically verify the individual fields on the metadata object, however the metadata is fetched and verified for most of them at some point during the testing. It might be something that I could add in another PR sometime (yay, test dev....)

The only item I had any concern about was the readability flags and I confirmed the logic for that in a separate project..

Reply:
> The only item I had any concern about was the readability flags and I confirmed the logic for that in a separate project..

This was the reason I had asked the question as well :) Works for me if you have verified it.
Put together a test PR for this area to improve the tests. That would be great.